src/includes/dinit-iostream.h

```diff
@@ -0,0 +1,738 @@
+ *      try {
+ *          std::unique_ptr<streambuf> my_buffer(new streambuf);
+ *      }
+ *      catch(const std::bad_alloc &) {
+ *          print_err("Cannot allocate memory for the output buffer!\n");
+ *          return -1;
+ *      }
+ *      dio::ostream output_b(std::move(my_buffer)); // Takes ownership of my_buffer
+ *
+ *      // Construction with a file descriptor number
+ *      dio::ostream output_c(3);
+ *      if (output_c.buffer_failure()) return -1; // Failure in buffer allocation
+ *
+ *      // Construction with a combination of a file descriptor and moving a
+ *      // pre-allocated unique_ptr of the buffer
+ *      dio::ostream output_d(3, std::move(buffer_2));
+ *
+ * Setting (or changing) the file descriptor after construction is available through the set_fd()
+ * function (See note in top of set_fd() func). opening and closing files is possible through open()
+ * function.
+ *
+ * Copy/move constructors/assignments and destructor
+ * -------------------------------------------------
+ *
+ * ostream objects cannot be moved, copied or assigned.
+ *
+ * The ostream destructor will write all of the buffer content into its file descriptor through the
+ * flush_nx() call and close the its current fd.
+ *
+ * Note: Failures in ostream destructor are impossible to catch.
+ *
+ * Public member inherited from io_base: is_open()
+ * ------------------------------------------------------
+ *
+ * ostream through base io_base class provides is_open() function which return true when current
+ * file path is open or false when it's not.
+ *
+ * Error handling and Stream states
+ * --------------------------------
+ *
+ * There are two possible error states in an ostream: buffer failure and I/O failure. These are each
+ * represented by a bit in the value returned by the current_state() function.
+ *
+ * 1. buffer_failbit is false by default and means the class constructor failed to allocate a unique_ptr
+ *    of the buffer at construction time.
+ * 2. io_failbit is false by default and means there is something wrong with system I/O functions.
+ *
+ * Note: if the io_failbit is set in the return value of current_state(), then io_failure() will
+ * return the errno value corresponding to the failed operation. If it is clear, io_failure() will return 0.
+ *
+ * Exception handling
+ * ------------------
+ *
+ * See the description for each function to know what exceptions can be thrown by functions.
+ *
+ * All exception-throwing functions have exception-free variants marked by the "_nx" suffix. Those
+ * functions guarantee not to throw any exceptions.
+ */
+class ostream : public io_base
+{
+    // io_error has different functionality, it will be set to errno which returned by system
+    // functions (such as read()).
+    int io_error = 0;
+
+    // Internal function to appending into the buffer.
+    ssize_t put(const char *msg, size_t count) noexcept;
+
+    // Throw exception based on current state
```

Comment:
mention what the `states` parameter is for

Reply:
Wasn't addressed?

Reply:
Is this the right fix: https://github.com/davmac314/dinit/pull/263/commits/135c5fa570289bc52d32e3f8bf33dfa8ab0a4648 ?

Reply:
No, part of the point of documentation on a function is to explain to someone who is writing code which calls the function what they can/should pass as arguments.

The comment "states parameter is for throwing appropriate exceptions" doesn't help. It doesn't explain what the value should be.

It's a bitmask right? So mention that fact, and say where the possible bit values come from (the `io_states` enum? and are all of them meaningful in this case or not?). Give enough information so that nobody has to look at the code of the function to figure out how they should call it.